src/NUnitFramework/framework/Internal/Filters/InFilter.cs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using NUnit.Framework.Interfaces;
+
+namespace NUnit.Framework.Internal.Filters
+{
+    /// <summary>
+    /// Optimized filter to check in condition using HashSet.
+    /// </summary>
+    internal sealed class InFilter : TestFilter
+    {
+        private readonly Func<ITest, string?> _selector;
+        private readonly HashSet<string?> _values;
+
+        /// <summary>
+        /// Constructs new InFilter.
+        /// </summary>
+        /// <param name="selector">Selector to get value to check.</param>
+        /// <param name="values">Valid values for the filter.</param>
+        public InFilter(Func<ITest, string?> selector, IEnumerable<string?> values)
+        {
+            _selector = selector;
+            _values = new HashSet<string?>(values);
+        }
+
+        /// <summary>
+        /// Tries to optimize OrFilter into InFilter if contained filters are all same and non-regex.
+        /// </summary>
+        public static bool TryOptimize(OrFilter orFilter, [NotNullWhen(true)] out InFilter? optimized)
+        {
+            // check if we have all filters accessing same field in OR fashion without regex
+            var hashCommonTypeAndNonRegex = orFilter.Filters.Count > 0;
+            Type? commonType = null;
+
+            // eagerly build expected value collection
+            var expectedValues = new List<string?>(orFilter.Filters.Count);
+
+            // use for for faster traversal without boxed enumerator
+            for (var i = 0; i < orFilter.Filters.Count; i++)
+            {
+                var filter = orFilter.Filters[i];
+
+                // make sure invariants are valid
+                commonType ??= filter.GetType();
```

Comment:
Would it make any difference if we do this initialization outside the loop ? E.g. remove this line and change the declaration to
```
Type? commonType = hashCommonTypeAndNonRegex ? orFilter.Filters[0] : null;
```

Reply:
That would change the semantics, the next line ensure that every filter in list has the same type, which is a requirement.

`hashCommonTypeAndNonRegex &= commonType == filter.GetType();`

Reply:
Isn't the semantics the same?

In both cases `commonType` will be the type of the first element in `orFilter.Filters`. I was just wondering if it would make any difference to set `commonType` outside the loop rather than in each iteration checking if `commonType` is `null` and in that case assign it to the type of the current filter (meaning that this will only be set in the first iteration, but we will check in each iteration)

Reply:
This was mostly a question as I was curious whether it made any difference (my feeling is that it don't, so I don't think this should hold back this PR). Thanks for the great work @lahma :+1 and thanks for having patience with us :)